storage/src/main/java/org/apache/kafka/storage/internals/log/LogConfig.java

```diff
@@ -500,22 +500,29 @@ public static void validateBrokerLogConfigValues(Map<?, ?> props,
      * The default values should be extracted from the KafkaConfig.
      * @param props The properties to be validated
      */
-    private static void validateTopicLogConfigValues(Map<?, ?> props,
-                                                     boolean isRemoteLogStorageSystemEnabled) {
+    public static void validateTopicLogConfigValues(Map<?, ?> props,
+                                                    boolean isRemoteLogStorageSystemEnabled,
+                                                    boolean isReceivingConfigFromStore) {
         validateValues(props);
         boolean isRemoteLogStorageEnabled = (Boolean) props.get(TopicConfig.REMOTE_LOG_STORAGE_ENABLE_CONFIG);
         if (isRemoteLogStorageEnabled) {
-            validateRemoteStorageOnlyIfSystemEnabled(isRemoteLogStorageSystemEnabled);
-            validateNoRemoteStorageForCompactedTopic(props);
-            validateRemoteStorageRetentionSize(props);
-            validateRemoteStorageRetentionTime(props);
+            validateRemoteStorageOnlyIfSystemEnabled(isRemoteLogStorageSystemEnabled, isReceivingConfigFromStore);
+            if (!isReceivingConfigFromStore) {
+                validateNoRemoteStorageForCompactedTopic(props);
+                validateRemoteStorageRetentionSize(props);
+                validateRemoteStorageRetentionTime(props);
+            }
         }
     }
 
-    private static void validateRemoteStorageOnlyIfSystemEnabled(boolean isRemoteLogStorageSystemEnabled) {
+    private static void validateRemoteStorageOnlyIfSystemEnabled(boolean isRemoteLogStorageSystemEnabled, boolean isReceivingConfigFromStore) {
         if (!isRemoteLogStorageSystemEnabled) {
-            throw new ConfigException("Tiered Storage functionality is disabled in the broker. " +
-                    "Topic cannot be configured with remote log storage.");
+            if (isReceivingConfigFromStore) {
```

Comment:
Can we update the message such that it's same for both the cases?

If the user disables the tiered storage functionality in the broker and restarts the node, and if any topic is left with remote storage enabled, then the same error message can be reused.

Reply:
What would your suggestion for such a message be? I left it as two separate ones because in one situation we say "hey, you cannot disable TS on the cluster while you have TS topics" while in the other we say "hey, you cannot enable TS on topics while you do not have TS on the cluster", but I see value in your suggestion. Would you like "You cannot have topics with enabled remote storage if Tiered Storage functionality is disabled"?

Reply:
The intention is not to supply the `boolean` to differentiate the cases. IMO, the existing error message should be sufficient:

> Tiered Storage functionality is disabled in the broker. Topics cannot be configured with remote log storage.
          
Let's wait for others opinion. cc @showuon @divijvaidya

Reply:
I think having 2 error messages make it more clear. After all, in the restart case, the user doesn't "configure" anything but seeing this message: `Topics cannot be configured with remote log storage.` is weird.

Reply:
Okay, I have kept the two messages. If there are more in favour of merging them together I can do this.